### PR TITLE
Add `wit_import` attribute macro to import functions from Wasm modules

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

The new Witty crate should allow calling functions exported from Wasm guest modules.

## Proposal

Use a `wit_import` attribute macro that parses a `trait` definition and generates the code to call functions from a guest Wasm instance defined by the trait.

In order to make the behavior independent from the host runtime, some traits were defined to abstract away the exported function's signature and how parameters and results are converted to the necessary WIT primitives to call the function.

## Test Plan

Some simple tests were written to ensure that the `wit_import` macro generates code that tries to load functions from a mock Wasm instance and allows calling them.

The `test_case` crate together with the `TestInstanceFactory` trait were used in order to make it simple for future runtime implementations to reuse the same tests.

## Links

Part of #906

## Release Plan

- [x] All good!
- [ ] Need to bump the major/minor version number in the next release of the crates.
- [ ] Need to update the developer manual.
- [ ] This PR is adding or removing Cargo features.
- [ ] Release is blocked and/or tracked by other issues (see links above)

## Reviewer Checklist

- [ ] The title and the summary of the PR are short and descriptive.
- [ ] The proposed solution achieves the goals stated in the PR.
- [ ] The test plan is reproducible and provides sufficient coverage.
- [ ] The release plan is adequate.
- [ ] The commits correspond to distinct logical changes.
- [ ] The code follows the [coding guidelines](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md).
- [ ] The proposed changes look correct.
- [ ] The CI is passing.
- [x] All of the above!

### CODE REVIEW

File: linera-witty/src/imported_function_interface/mod.rs [hunk 1]
```diff
@@ -12,3 +12,83 @@
...
+    /// The type representing the host-side results.
+    type HostResults: WitLoad;
+
+    /// The flat layout representing the guest-side parameters.
+    type GuestParameters: FlatLayout;
+
+    /// The flat layout representing the guest-side results.
+    type GuestResults: FlatLayout;
+
+    /// Converts the the host-side parameters into the guest-side parameters.
```
Comment (ma2bd) [inline linera-witty/src/imported_function_interface/mod.rs:39]: typo: the the

File: linera-witty/src/imported_function_interface/mod.rs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    fn lift_results<Instance>(
+        results: Self::GuestResults,
+        memory: &Memory<'_, Instance>,
+    ) -> Result<Self::HostResults, RuntimeError>
+    where
+        Instance: InstanceWithMemory,
+        <Instance::Runtime as Runtime>::Memory: RuntimeMemory<Instance>;
+}
+
+impl<Parameters, Results> ImportedFunctionInterface for (Parameters, Results)
```
Comment (ma2bd) [inline linera-witty/src/imported_function_interface/mod.rs:58]: Is this the only implementation of `ImportedFunctionInterface`? Did you consider a helper struct `struct ImportedFunctionInterface<P, R> { parameters: P, results: R }`, then defining a block `impl<P, R> ImportedFunctionInterface<P, R> where .. { .. }` ?

Comment (ma2bd) [reply linera-witty/src/imported_function_interface/mod.rs:58]: I guess it depends if you need the output associated types `Guest*` as a short alias.

File: linera-witty/tests/common/test_instance.rs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+/// A factory of [`MockInstance`]s.
+#[derive(Default)]
+pub struct MockInstanceFactory {
+    deferred_assertions: Vec<Box<dyn Any>>,
+}
+
+impl TestInstanceFactory for MockInstanceFactory {
+    type Instance = MockInstance;
+
+    fn load_test_module(&mut self, module: &str) -> Self::Instance {
```
Comment (ma2bd) [inline linera-witty/tests/common/test_instance.rs:26]: The tests are really cool!

**Review by ma2bd — APPROVED** [review]
LGTM!

File: linera-witty/src/imported_function_interface/mod.rs (hunk 2 above)
Comment (jvff) [reply linera-witty/src/imported_function_interface/mod.rs:58]: Yes, unfortunately associated types in types isn't stable [yet](https://github.com/rust-lang/rust/issues/8995) :/

File: linera-witty/src/imported_function_interface/mod.rs (hunk 1 above)
Comment (jvff) [reply linera-witty/src/imported_function_interface/mod.rs]: Done, thanks!

**Pull request merged**